synapse/handlers/directory.py

```diff
@@ -525,7 +526,7 @@ async def edit_published_appservice_room_list(
 
     async def get_aliases_for_room(
         self, requester: Requester, room_id: str
-    ) -> List[str]:
+    ) -> Sequence[str]:
```

Comment:
I recommend using `Collection[str]` here if possible since `str` is a `Sequence[str]`, so it could let bugs slip in.

Reply:
(Looks like this might apply in a few places below.)

Reply:
`str` is also a subtype of `Collection[str]`, I think?

Reply:
```
$ cat temp.py 
from typing import Collection
x: Collection[str]
x = "123"

$ mypy --strict temp.py 
Success: no issues found in 1 source file
```

Reply:
Right -- we need a `FrozenList` sort of class.

Reply:
For cross-referencing, https://github.com/python/typing/issues/256. https://github.com/python/typing/issues/510 and https://github.com/python/typing/issues/256#issuecomment-917601678 are interesting ideas though.

Reply:
https://stackoverflow.com/questions/11142397/does-python-have-an-immutable-list suggests that tuples can be considered to be `FrozenList`s. We could copy all these lists into tuples? (which I'm not a fan of)

Alternatively it may be possible to define some sort of `ReadOnlyList[T]` protocol which is somehow only satisfied by a real list (`__class__` has to be `Type[List[T]]` maybe?) but doesn't expose any methods to mutate it. However, that doesn't feel very python-y.

Any thoughts on which way we should go?

Reply:
Just my opinions in passing:

- I don't mind tuples as frozen sequence types! But I imagine we'd end up spending a non-negligible chunk of time converting lists to tuples
- Bluntly: I wouldn't block this on trying to fix/improve Python's type system.
- _Maybe_ we could do something clever with a mypy plugin, but I think that's [a rabbit hole](https://youtu.be/rVJHQ7SkgwI?t=3).

Reply:
> _Maybe_ we could do something clever with a mypy plugin, but I think that's [a rabbit hole](https://youtu.be/rVJHQ7SkgwI?t=3).

I treated myself. See https://github.com/matrix-org/synapse/pull/14911

Reply:
Since this comment we now have `StrCollection` as our home-grown workaround for this problem.

I think we should punt this problem to https://github.com/matrix-org/synapse/issues/14809 and try to land this PR as-is.